### ISSUE TITLE
kubernetes (k8s) connection stops working after switching namespace, kube config changes cluster name, error x509: certificate signed by unknown authority

### ISSUE THREAD

**Issue by b2cc** [gravitational/teleport#22835]
**Description**
I'm currently facing the issue that when I log into kubernetes (OKD 4.12 in this case), the connection stops working after I switch into a different kubernetes namespace, and I get the error `Unable to connect to the server: x509: certificate signed by unknown authority` for every request. After some debugging I noticed that the `.kube/config` file had changed and the cluster context hat a seconf, wrong `cluster` entry that was causing this issues. 

I followed these docs to set up the k8s/teleport integration:
* https://goteleport.com/docs/kubernetes-access/register-clusters/static-kubeconfig/

I have attached all my configs the the relevant `.kube` files at the end of the report.

**Current behavior**
The initial login to teleport works, and I can see the k8s cluster:
```
~# tsh kube ls
Kube Cluster Name                                 Labels Selected 
------------------------------------------------- ------ -------- 
fedi-okd                                                          
```

Log in to the cluster works as well:
```
~# tsh kube login fedi-okd
Logged into Kubernetes cluster "fedi-okd". Try 'kubectl version' to test the connection.

~# oc whoami
<username>

~# oc version
Client Version: 4.12.0-0.okd-2023-02-18-033438
Kustomize Version: v4.5.7
Kubernetes Version: v1.25.0-2653+a34b9e9499e6c3-dirty
```

It works correctly, because I can list pods in projects that are assigned to the user (as long as I don't switch the namespace):
```
~# oc get pod -n <namespace>
NAME                         READY   STATUS    RESTARTS   AGE
pod-abcdef-8f975b649-lkf98   1/1     Running   0          7d17h
```

Cluster context looks ok as well:
```
~# oc config get-contexts
CURRENT   NAME                            CLUSTER                AUTHINFO                        NAMESPACE
*         teleport.example.com-fedi-okd   teleport.example.com   teleport.example.com-fedi-okd
```

Problems start when I switch into a different namespace:
```
oc project <project>
Now using project "<project>" on server "https://teleport.example.com:443".
```

Now I have a second context which has wrong details (notice how the `CLUSTER` name now has a dash instead of a period, and the additional `:443` at the end) and from here on I get the aforementioned certificate errors:
```
~# oc config get-contexts
CURRENT   NAME                                             CLUSTER                    AUTHINFO                               NAMESPACE
*         <namespace>/teleport-example.com:443/<username>  teleport-example.com:443   <username>/teleport-example.com:443    <namespace>
          teleport.example.com-fedi-okd                    teleport.example.com       teleport.example.com-fedi-okd
```

Error:
```
~# oc whoami
Unable to connect to the server: x509: certificate signed by unknown authority
```

I checked all my configs and logs, but I couldn't find anything what would mess with the `kube/config` in such a way.
What could be the reason for this behavior? Am I missing some configuration?

**Bug details:**

teleport version:
```
Teleport v12.0.4 git:api/v12.0.4-0-gcf72ee5d59 go1.20.1
```

`teleport.yaml` on teleport server:
```
version: v3
teleport:
  nodename: teleport
  data_dir: /var/lib/teleport
  log:
    output: stderr
    severity: INFO
    format:
      output: text
  ca_pin: ""
  diag_addr: ""
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  cluster_name: teleport.example.com
  proxy_listener_mode: multiplex
ssh_service:
  enabled: "yes"
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:443
  public_addr: teleport.example.com:443
  kube_listen_addr: 0.0.0.0:3026
  kube_public_addr: teleport.example.com:3026
  https_keypairs: []
  https_keypairs_reload_interval: 0s
  acme:
    enabled: "yes"
    email: b2c@example.com
kubernetes_service:
  enabled: "yes"
  kubeconfig_file: "/etc/teleport.d/kubeconfig"
  listen_addr: 0.0.0.0:3027
```

`kubeconfig` on teleport server:
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0 ... S0K
    server: https://api.okd.in.example.com:6443
  name: fedi-okd
contexts:
- context:
    cluster: fedi-okd
    user: fedi-okd-teleport-sa
  name: fedi-okd
current-context: fedi-okd
kind: Config
preferences: {}
users:
- name: fedi-okd-teleport-sa
  user:
    token: eyJhb ... xM
```

Initial `.kube/config` on client after login (connection working):
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0 ... tLS0K
    server: https://teleport.example.com:443
    tls-server-name: kube.teleport.example.com
  name: teleport.example.com
contexts:
- context:
    cluster: teleport.example.com
    user: teleport.example.com-fedi-okd
  name: teleport.example.com-fedi-okd
current-context: teleport.example.com-fedi-okd
kind: Config
preferences: {}
users:
- name: teleport.example.com-fedi-okd
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - kube
      - credentials
      - --kube-cluster=fedi-okd
      - --teleport-cluster=teleport.example.com
      command: /opt/teleport/teleport-12.0.4/tsh
      env: null
      provideClusterInfo: false
```

`.kube/config` on client after namespace change (connection stops working, x509 error, referenced certificate is the same in both clusters/contexts):
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0 ... S0K
    server: https://teleport.example.com:443
  name: teleport-dest-unreachable-net:443
- cluster:
    certificate-authority-data: LS0 ... S0K
    server: https://teleport.example.com:443
    tls-server-name: kube.teleport.example.com
  name: teleport.example.com
contexts:
- context:
    cluster: teleport-dest-unreachable-net:443
    namespace: mastodon-wien-rocks
    user: b2c/teleport-dest-unreachable-net:443
  name: mastodon-wien-rocks/teleport-dest-unreachable-net:443/b2c
- context:
    cluster: teleport.example.com
    user: teleport.example.com-fedi-okd
  name: teleport.example.com-fedi-okd
current-context: mastodon-wien-rocks/teleport-dest-unreachable-net:443/b2c
kind: Config
preferences: {}
users:
- name: b2c/teleport-dest-unreachable-net:443
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - kube
      - credentials
      - --kube-cluster=fedi-okd
      - --teleport-cluster=teleport.example.com
      command: /opt/teleport/teleport-12.0.4/tsh
      env: null
      interactiveMode: IfAvailable
      provideClusterInfo: false
- name: teleport.example.com-fedi-okd
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - kube
      - credentials
      - --kube-cluster=fedi-okd
      - --teleport-cluster=teleport.example.com
      command: /opt/teleport/teleport-12.0.4/tsh
      env: null
      interactiveMode: IfAvailable
      provideClusterInfo: false
```

**Comment by b2cc** [comment]
Ok I think I found the issue: `tls-server-name: kube.teleport.example.com` is missing in the new entry that gets created by the `oc` binary. Was pulling my hair out over this.

A workaround would be to call `oc --tls-server-name='kube.teleport.example.com'` but that's getting tedious fast.
Does anyone know how this can be adjusted to match up or what needs to be configured to make Openshift work with teleport in such a setup?

**Comment by strideynet** [comment]
I'm fairly convinced this is poor behaviour of the `oc` client.

Potentially patched by OC in https://github.com/openshift/oc/commit/f47f44da57d408e4d9b0fd2fcaaf3f056d999913 - and released in 4.14. Can you test with 4.14 ?

**Comment by b2cc** [comment]
@strideynet : nice find, I didn't even think of this anymore - also you're right, definitely not a teleport bug, but an Openshift problem. I just couldn't figure out what was the root cause for this in the beginning.
Just checked, and it works now with the latest 4.14 release, e.g. `openshift-client-linux-4.14.0-0.okd-2023-10-28-073550`.

Thanks!
